Implement sliding-window smoothing for production, consumption, and battery with battery recharge reserve

HAuser1234 opened the pull request:
This pull request introduces configurable sliding-window smoothing for solar production, consumption, and battery charge power readings, as well as a new option to reserve a portion of solar production for battery charging when the battery is not full. These changes make the optimizer more robust against short-term fluctuations and help prioritize battery charging when desired.

The most important changes include:

**Smoothing and Reserve Features:**

* Added configuration options for sliding-window smoothing of solar production, consumption, and battery charge power, with user-defined window lengths in minutes. These options allow the system to average readings over a specified period, reducing the impact of short-term spikes or drops. (`config_schema.py`, `const.py`, `coordinator.py`, `strings.json`) [[1]](diffhunk://#diff-884dc9e30aa06ad4192085fe4299916a484c7fea417def5442367b94d8ff2732R56-R59) [[2]](diffhunk://#diff-e9cd7ad5d7b83ed3e848245532a11539834c4d4f068b0af04969c471f5a7d283R22-R25) [[3]](diffhunk://#diff-e9cd7ad5d7b83ed3e848245532a11539834c4d4f068b0af04969c471f5a7d283R62-R65) [[4]](diffhunk://#diff-c5d1f3c0a7cb4bb930f725a3e818ced281be9eb3ab87353365f0e000824fc5e2L26-R40) [[5]](diffhunk://#diff-c5d1f3c0a7cb4bb930f725a3e818ced281be9eb3ab87353365f0e000824fc5e2R81-R90) [[6]](diffhunk://#diff-c5d1f3c0a7cb4bb930f725a3e818ced281be9eb3ab87353365f0e000824fc5e2R146-R153) [[7]](diffhunk://#diff-c5d1f3c0a7cb4bb930f725a3e818ced281be9eb3ab87353365f0e000824fc5e2L158-R263) [[8]](diffhunk://#diff-0a125b3720a6f6aad31fdc550ed6983fd9979579303296c26383c60fa3ecb669L30-R34) [[9]](diffhunk://#diff-0a125b3720a6f6aad31fdc550ed6983fd9979579303296c26383c60fa3ecb669L43-R51) [[10]](diffhunk://#diff-0a125b3720a6f6aad31fdc550ed6983fd9979579303296c26383c60fa3ecb669L162-R174) [[11]](diffhunk://#diff-0a125b3720a6f6aad31fdc550ed6983fd9979579303296c26383c60fa3ecb669L175-R191)
* Implemented a new configuration option to reserve a specified number of watts from solar production for battery charging when the battery state of charge is below 100%. This ensures battery charging is prioritized over other loads when needed. (`config_schema.py`, `const.py`, `coordinator.py`, `strings.json`) [[1]](diffhunk://#diff-884dc9e30aa06ad4192085fe4299916a484c7fea417def5442367b94d8ff2732R56-R59) [[2]](diffhunk://#diff-e9cd7ad5d7b83ed3e848245532a11539834c4d4f068b0af04969c471f5a7d283R22-R25) [[3]](diffhunk://#diff-e9cd7ad5d7b83ed3e848245532a11539834c4d4f068b0af04969c471f5a7d283R62-R65) [[4]](diffhunk://#diff-c5d1f3c0a7cb4bb930f725a3e818ced281be9eb3ab87353365f0e000824fc5e2L26-R40) [[5]](diffhunk://#diff-c5d1f3c0a7cb4bb930f725a3e818ced281be9eb3ab87353365f0e000824fc5e2R81-R90) [[6]](diffhunk://#diff-c5d1f3c0a7cb4bb930f725a3e818ced281be9eb3ab87353365f0e000824fc5e2R146-R153) [[7]](diffhunk://#diff-c5d1f3c0a7cb4bb930f725a3e818ced281be9eb3ab87353365f0e000824fc5e2L187-R313) [[8]](diffhunk://#diff-0a125b3720a6f6aad31fdc550ed6983fd9979579303296c26383c60fa3ecb669L30-R34) [[9]](diffhunk://#diff-0a125b3720a6f6aad31fdc550ed6983fd9979579303296c26383c60fa3ecb669L43-R51) [[10]](diffhunk://#diff-0a125b3720a6f6aad31fdc550ed6983fd9979579303296c26383c60fa3ecb669L162-R174) [[11]](diffhunk://#diff-0a125b3720a6f6aad31fdc550ed6983fd9979579303296c26383c60fa3ecb669L175-R191)

**Implementation and Code Structure:**

* Refactored the coordinator to use a new `_apply_smoothing_window` method, which manages the sliding window for each smoothed value using a `deque`, and updated the update logic to use this method for production, consumption, and battery charge power. (`coordinator.py`) [[1]](diffhunk://#diff-c5d1f3c0a7cb4bb930f725a3e818ced281be9eb3ab87353365f0e000824fc5e2R5) [[2]](diffhunk://#diff-c5d1f3c0a7cb4bb930f725a3e818ced281be9eb3ab87353365f0e000824fc5e2R81-R90) [[3]](diffhunk://#diff-c5d1f3c0a7cb4bb930f725a3e818ced281be9eb3ab87353365f0e000824fc5e2R164-R204) [[4]](diffhunk://#diff-c5d1f3c0a7cb4bb930f725a3e818ced281be9eb3ab87353365f0e000824fc5e2L158-R263) [[5]](diffhunk://#diff-c5d1f3c0a7cb4bb930f725a3e818ced281be9eb3ab87353365f0e000824fc5e2L187-R313)
* Updated the configuration schema, constants, and localization files to support the new options, including default values, user-facing strings, and descriptions for each new setting. (`config_schema.py`, `const.py`, `strings.json`) [[1]](diffhunk://#diff-884dc9e30aa06ad4192085fe4299916a484c7fea417def5442367b94d8ff2732R56-R59) [[2]](diffhunk://#diff-e9cd7ad5d7b83ed3e848245532a11539834c4d4f068b0af04969c471f5a7d283R22-R25) [[3]](diffhunk://#diff-e9cd7ad5d7b83ed3e848245532a11539834c4d4f068b0af04969c471f5a7d283R62-R65) [[4]](diffhunk://#diff-0a125b3720a6f6aad31fdc550ed6983fd9979579303296c26383c60fa3ecb669L30-R34) [[5]](diffhunk://#diff-0a125b3720a6f6aad31fdc550ed6983fd9979579303296c26383c60fa3ecb669L43-R51) [[6]](diffhunk://#diff-0a125b3720a6f6aad31fdc550ed6983fd9979579303296c26383c60fa3ecb669L162-R174) [[7]](diffhunk://#diff-0a125b3720a6f6aad31fdc550ed6983fd9979579303296c26383c60fa3ecb669L175-R191)
*  Added german translation

These enhancements provide users with more control over data smoothing and battery management, improving the stability and flexibility of the solar optimizer.

Copilot committed:
Initial plan

Copilot committed:
Implement sliding-window smoothing and battery reserve features

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Add separate smoothing windows for consumption and battery power

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Fix naming convention and add translations for new configuration options

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

HAuser1234 committed:
Implement sliding-window smoothing for production, consumption, and battery with battery recharge reserve

Implement sliding-window smoothing for production, consumption, and battery with battery recharge reserve

HAuser1234 commented:
<img width="1715" height="325" alt="image" src="https://github.com/user-attachments/assets/501dbc5a-f9cb-4a5b-bbd8-a933d077f768" />
It seems to be able to hold a steady battery charging offset (when battery < 100%) and helping to ignore temporary swings in production helping the optimizer to get a more persistant solution.

Copilot committed:
Initial plan

Copilot committed:
Add battery recharge reserve before smoothing option

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Add tests for battery reserve before/after smoothing

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Add translations for battery reserve before smoothing option

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Update translations to include new option in en.json and shorten descriptions

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Add French translations and battery reserve reduction attribute to power_production sensor

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

HAuser1234 committed:
Add battery recharge reserve option in German translation

HAuser1234 committed:
Merge pull request #2 from HAuser1234/copilot/add-switch-for-battery-recharge

Add option to apply battery reserve before smoothing

HAuser1234 committed:
Fix formatting in German translation for solar optimizer

HAuser1234 committed:
Add translation for battery recharge reserve before smoothing

Added a new translation key for battery recharge reserve before smoothing.

Copilot committed:
Initial plan

Copilot committed:
Update core algorithm and config to use household consumption

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Update README files with simplified power distribution documentation

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Fix algorithm to explicitly account for production in net calculation

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Address code review comments - fix spelling and language consistency

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Fix household consumption calculation and add diagnostic sensors

- Calculate base household by subtracting currently distributed power from total consumption
- Add protection against negative values during smoothing transients
- Add diagnostic sensors: household_consumption, available_excess_power, total_current_distributed_power
- Update algorithm comments to clarify household_consumption excludes managed devices

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Fix algorithm to use direct power sum instead of diff

The algorithm was incorrectly using diff from initial power, causing it to think there was excess power when devices were already active. Since household_consumption already excludes currently active managed devices, we should add solution devices directly, not the difference.

Example: Wallbox at 4830W, household base 670W, production 3000W
- OLD (buggy): total = 670 + (4830 - 4830) = 670W → thinks exporting 2330W
- NEW (fixed): total = 670 + 4830 = 5500W → correctly shows importing 2500W

This prevents the algorithm from increasing device power when already importing from grid.

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Add household smoothing and min export margin features

1. Add household consumption smoothing (smoothing_household_window_min):
   - Smooths only base household consumption (excluding managed devices)
   - Helps compensate for short spikes from fridges, kettles, etc.
   - Never allows base consumption to go negative

2. Add minimum export margin at 100% SOC (min_export_margin_w):
   - Reserves watts from optimization when battery is full
   - Positive values keep battery topped up by always exporting
   - Negative values allow slight import
   - Prevents battery discharge when at 100%

Updated config schema and all translations (en, fr, de, strings.json) with descriptions for both new features.

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Fix power deficit handling and remove battery smoothing

1. Fix available excess power calculation when deficit occurs:
   - When household_consumption - devices goes negative (power deficit due to reporting lag)
   - Set available_excess to 0 instead of jumping to full PV production
   - Add explicit deficit detection and logging

2. Improve household smoothing:
   - Add extra protection to ensure smoothed value stays non-negative
   - Properly track raw vs smoothed values

3. Remove all battery smoothing references:
   - Remove CONF_SMOOTHING_BATTERY_WINDOW_MIN from const.py
   - Remove from config_schema.py
   - Remove battery_window deque and related code from coordinator.py
   - Battery charge power now recorded for diagnostics only (no smoothing)

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

HAuser1234 committed:
Merge pull request #3 from HAuser1234/copilot/implement-simplified-power-distribution

Refactor power distribution to use household consumption with enhanced battery management

Copilot committed:
Initial plan

Copilot committed:
Fix power distribution bug caused by HA state lag

When devices are turned on/off, there's a lag between sending the command and HA state updating. The algorithm was reading device states that hadn't updated yet, causing it to underestimate power consumption and turn on too many devices.

Fix: Use requested_power (from previous cycle) when it's higher than actual state power, to account for state lag during device transitions. This prevents the algorithm from thinking a newly-activated device is using 0W.

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Add documentation for state lag handling in coordinator

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Address code review feedback - improve comment clarity

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Fix deactivation lag handling - complete state lag fix

The original fix only handled activation lag (device turned on but HA shows off).
Now also handles deactivation lag (device turned off but HA shows on).

This prevents the algorithm from counting a just-deactivated device as still
using power, which was causing it to turn on multiple devices simultaneously
when only one could be supported by available power.

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Address code review feedback - improve error handling and testing

- Fix service parsing check: len(parties) < 3 (was < 2) to match 'domain/action/attribute' format
- Add safe handling for non-numeric/absent sensor values with try/except
- Change lag detection logging from INFO to DEBUG to reduce noise
- Use .get() for attributes dictionary access to avoid KeyError
- Add comprehensive multi-device integration test for 750W/950W scenario
- Remove flaky Test 5 and focus on core state lag scenarios

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

HAuser1234 committed:
Merge pull request #5 from HAuser1234/copilot/investigate-power-distribution-bug

Fix device power tracking during HA state lag causing overconsumption

Copilot committed:
Initial plan

Copilot committed:
Add device switching stability to prevent excessive device changes

- Add switching_penalty_factor parameter to algorithm (default: 0.5)
- Penalize solutions that turn off currently active devices
- Penalty is proportional to device power and inversely related to benefit
- Encourages incremental device addition vs. swapping
- Reduces relay wear and ensures devices reach minimum on-time
- Configurable via solar_optimizer.yaml algorithm section
- Added tests for switching stability behavior
- Updated README with documentation

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Clean up code formatting (remove trailing whitespace, fix line length)

- Remove trailing whitespace from switching penalty code
- Break long lines for better readability
- Improve code formatting to meet pylint standards
- Increased pylint score from 8.02 to 8.52

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Fix code review issues: avoid code duplication and NameError

- Extract penalty calculation to single variable (penalty_value)
- Initialize penalty_value before conditional to avoid NameError
- Eliminate code duplication in debug logging
- Improve code clarity and maintainability

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Move switching_penalty_factor to UI config and clarify variable power device behavior

- Add CONF_SWITCHING_PENALTY_FACTOR constant with default 0.5
- Add switching_penalty_factor to central_config_schema for UI configuration
- Update coordinator to read from config entry and apply to algorithm
- Add UI strings and descriptions for the new parameter
- Clarify in algorithm comments: variable power devices can change power freely, only OFF is penalized
- Update README: document UI configuration, deprecate YAML config
- Maintain backward compatibility with YAML algorithm config

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Fix missing import and remove YAML backward compatibility

- Add DEFAULT_SWITCHING_PENALTY_FACTOR and CONF_SWITCHING_PENALTY_FACTOR imports to coordinator.py
- Remove YAML backward compatibility code (switching_penalty_factor no longer read from solar_optimizer.yaml)
- Update README to remove mentions of YAML backward compatibility
- Fixes NameError: name 'DEFAULT_SWITCHING_PENALTY_FACTOR' is not defined

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Add translations for switching_penalty_factor in EN, FR, and DE

- Added English translation: "Device switching penalty factor"
- Added French translation: "Facteur de pénalité de commutation des appareils"
- Added German translation: "Geräte-Schaltbestrafungsfaktor"
- Translations added to both config and options sections in all language files
- Includes descriptive text explaining the parameter in each language

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Fix config flow: Use NumberSelector for switching_penalty_factor

- Changed from vol.Coerce(float) to selector.NumberSelector widget
- Added proper UI controls: min=0.0, max=5.0, step=0.1, mode=BOX
- Ensures field appears correctly in Home Assistant config flow UI
- Allows users to adjust the value via slider/number input

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

jmcollin78 reviewed (commented):
Hello, thank you for this. There is already smoothing function in SO and battry threshold so it is not clear what this PR improve considering the existing features. I will have a look but I'm on a big change on the Versatile Thermostat and I will not be available soon.

HAuser1234 committed:
Merge pull request #6 from HAuser1234/copilot/optimize-power-switching-algorithm

Add switching penalty to reduce device churn from marginal power changes

HAuser1234 committed:
Revert "Fix device power tracking during HA state lag causing overconsumption"

HAuser1234 committed:
Merge pull request #8 from HAuser1234/revert-5-copilot/investigate-power-distribution-bug

Revert "Fix device power tracking during HA state lag causing overconsumption"

HAuser1234 committed:
Revert "Revert "Fix device power tracking during HA state lag causing overconsumption""

HAuser1234 committed:
Merge pull request #9 from HAuser1234/revert-8-revert-5-copilot/investigate-power-distribution-bug

Revert "Revert "Fix device power tracking during HA state lag causing overconsumption""

HAuser1234 committed:
Revert "Revert "Revert "Fix device power tracking during HA state lag causing overconsumption"""

HAuser1234 committed:
Merge pull request #10 from HAuser1234/revert-9-revert-8-revert-5-copilot/investigate-power-distribution-bug

Revert "Revert "Revert "Fix device power tracking during HA state lag causing overconsumption"""

Copilot committed:
Initial plan

Copilot committed:
Fix household deficit handling - compute excess correctly

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Apply Black formatting

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

HAuser1234 committed:
Merge pull request #12 from HAuser1234/copilot/fix-excess-power-calculation

Fix household deficit zeroing available excess power despite sufficient production

Copilot committed:
Initial plan

Copilot committed:
Implement standby handling and off-penalty + debounce for 0W devices

- Remove current_power <= 0 force-off condition in simulated_annealing_algo.py
- Add was_active tracking to equipment dict for initial device state
- Update switching penalty to use was_active instead of current_power > 0
- Add _last_non_zero_power tracking in managed_device.py
- Implement debounce in set_current_power_with_device_state to use requested_power or last_non_zero_power during grace window

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Add auto-suggest switching penalty and optional price clamping

- Add auto_switching_penalty option to automatically calculate optimal penalty based on conditions
- Add clamp_price_step option to reduce price volatility (e.g., 0.05 for 5-cent steps)
- Implement _calculate_optimal_switching_penalty method considering production, consumption, and active devices
- Implement _clamp_price method to round prices to configured steps
- Update algorithm initialization to accept new parameters

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Add suggested penalty sensor and update coordinator

- Add suggested_penalty property to algorithm to expose last calculated value
- Update coordinator to track and expose suggested penalty
- Add new sensor entity for suggested_penalty in sensor.py
- Configure sensor with appropriate icon, device_class, and state_class
- Update coordinator configure method to read and apply auto_switching_penalty and clamp_price_step from config

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Add translations for auto_switching_penalty and clamp_price_step

- Add English translations for new config options
- Add French translations for auto penalty and price clamping
- Add German translations for auto penalty and price clamping
- Translations include both field names and detailed descriptions

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Add config flow options and fix suggested_penalty sensor

- Add auto_switching_penalty boolean option to config schema
- Add clamp_price_step number option to config schema (0.0-1.0, step 0.01)
- Always calculate suggested penalty even when auto mode is off (for monitoring/tuning)
- This fixes the suggested_penalty sensor showing as unknown

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Add reward for turning on devices with excess power

- When a device was off but solution turns it on, check if there's still excess power
- If turning on the device would still leave excess power being exported, apply a reward (negative penalty)
- Reward is 50% of penalty factor to encourage using available solar without being too aggressive
- This solves the issue where devices stay off even with abundant excess power
- Devices will now be more likely to turn back on when plenty of solar is available

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

Copilot committed:
Add priority-aware reward boost for low-priority devices with excess power

- Low-priority devices (priority > 8) get additional reward boost when abundant excess power (>20%)
- Boost scales with both priority level (higher priority number = more boost) and excess ratio
- Boost is weighted by priority_weight to counteract the priority penalty that keeps them off
- This ensures low-priority devices turn on when there's plenty of solar available
- Addresses issue where devices with LOW priority stay off despite 5kW+ excess power

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

HAuser1234 committed:
Merge pull request #13 from HAuser1234/copilot/fix-device-power-issues

Fix standby/0W device handling, add auto-penalty, price clamping, and excess power rewards

Copilot committed:
Initial plan

Copilot committed:
Fix coordinator selector to show device type options after adding coordinator

Co-authored-by: HAuser1234 <122117318+HAuser1234@users.noreply.github.com>

HAuser1234 committed:
Merge pull request #14 from HAuser1234/copilot/fix-coordinator-selector-error

Fix config flow device type selector after coordinator setup

jmcollin78 commented on custom_components/solar_optimizer/simulated_annealing_algo.py:
```diff
@@ -205,54 +354,174 @@ def recuit_simule(
+        # if a device was running, which correctly handles standby/0W devices.
+        # For variable power devices, changing power (up or down) is NOT penalized.
+        # Only turning the device completely OFF/ON incurs a penalty/reward.
+        switching_penalty = 0
```
This block of code is very hard to read and understand. There is too much complexity (imbricated if).

jmcollin78 commented on custom_components/solar_optimizer/translations/fr.json:
```diff
@@ -27,20 +27,40 @@
                 "data_description": {
                     "refresh_period_sec": "Période de rafraichissement en secondes. Attention des calculs lourds sont effectués à chaque période. Ne pas rafraichir trop souvent",
-                    "power_consumption_entity_id": "l'entity_id du capteur de consommation nette. La consommation nette doit être négative si l'énergie est exportée vers le réseau.",
+                    "power_consumption_entity_id": "L'entity_id du capteur de consommation électrique de la maison (en watts positifs). Cela devrait représenter la consommation de base de la maison. Si vous utilisez un capteur de compteur net, envisagez de créer un capteur template qui fournit des valeurs de consommation positives.",
```
This is a major breaking change if I understand correctly. All current installation will no more work. What we need here: is the grid import/export net value. If you change this, the algorithm should change. I see you change also the algorithm.

Don't know what to think about that. Was it really necessary to change this ? I guess not.

jmcollin78 reviewed (changes requested):
You change too many things in one PR:
1. smoothing: seems fine,
2. change the meaning of power_consumption_entity_id (really needed ?)
3. add debounce (I understand this can be necessary),
4. switching penalty factor 

This is almost a new integration you made and therefore it is very hard to see all eventual impacts.
But this is really a hard work (I guess IA helps...).

What do you think ?
